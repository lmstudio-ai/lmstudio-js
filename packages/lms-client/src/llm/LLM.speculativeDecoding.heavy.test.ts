import { Chat, LMStudioClient, type LLM } from "../index.js";
import {
  ensureHeavyTestsEnvironment,
  llmTestingQwen05B,
  llmTestingQwen3B,
} from "../shared.heavy.test.js";

describe("LLM + speculative decoding", () => {
  let client: LMStudioClient;
  let model: LLM;
  const chat = Chat.from([
    { role: "system", content: "This is the system prompt." },
    { role: "user", content: "User message 1" },
    { role: "assistant", content: "Assistant message 1" },
    { role: "user", content: "User message 2" },
  ]);
  beforeAll(async () => {
    client = new LMStudioClient();
    await ensureHeavyTestsEnvironment(client);
  });
  beforeEach(async () => {
    model = await client.llm.model(llmTestingQwen3B, {
      verbose: false,
      config: {
        llamaKCacheQuantizationType: "f32",
        llamaVCacheQuantizationType: "f32",
      },
    });
  }, 60_000);
  it("should work with .respond", async () => {
    const result = await model.respond(chat, {
      temperature: 0,
      draftModel: llmTestingQwen05B,
    });
    expect(result.content).toMatchSnapshot();
    expect(result.stats).toMatchSnapshot({
      numGpuLayers: expect.any(Number),
      timeToFirstTokenSec: expect.any(Number),
      tokensPerSecond: expect.any(Number),
      usedDraftModelKey: expect.any(String),
    });
    expect(result.modelInfo).toMatchSnapshot({
      identifier: expect.any(String),
      instanceReference: expect.any(String),
      modelKey: expect.any(String),
    });
    expect(result.roundIndex).toEqual(0);
  });
  it("should work with .complete", async () => {
    const result = await model.complete("1 + 1 = 2; 2 + 2 = ", {
      temperature: 0,
      maxTokens: 5,
      stopStrings: ["+"],
      draftModel: llmTestingQwen05B,
    });
    expect(result.content).toMatchInlineSnapshot(`"4; 3"`);
    expect(result.stats).toMatchSnapshot({
      numGpuLayers: expect.any(Number),
      timeToFirstTokenSec: expect.any(Number),
      tokensPerSecond: expect.any(Number),
      usedDraftModelKey: expect.any(String),
    });
    expect(result.modelInfo).toMatchSnapshot({
      identifier: expect.any(String),
      instanceReference: expect.any(String),
      modelKey: expect.any(String),
    });
    expect(result.roundIndex).toEqual(0);
  });
  it("should fail with invalid model", async () => {
    await expect(async () => {
      await model.complete("1 + 1 = 2; 2 + 2 = ", {
        draftModel: "invalid-model",
      });
    }).rejects.toThrow('Cannot find model "invalid-model"');
  });
});
