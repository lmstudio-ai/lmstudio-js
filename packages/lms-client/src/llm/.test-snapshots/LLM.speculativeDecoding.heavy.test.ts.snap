// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LLM + speculative decoding should work with .complete 2`] = `
{
  "acceptedDraftTokensCount": 2,
  "ignoredDraftTokensCount": 0,
  "numGpuLayers": Any<Number>,
  "predictedTokensCount": 5,
  "promptTokensCount": 15,
  "rejectedDraftTokensCount": 1,
  "stopReason": "maxPredictedTokensReached",
  "timeToFirstTokenSec": Any<Number>,
  "tokensPerSecond": Any<Number>,
  "totalDraftTokensCount": 3,
  "totalTokensCount": 20,
  "usedDraftModelKey": Any<String>,
}
`;

exports[`LLM + speculative decoding should work with .complete 3`] = `
{
  "architecture": "qwen2",
  "contextLength": 4096,
  "displayName": "Qwen2.5 3B Instruct",
  "format": "gguf",
  "identifier": Any<String>,
  "instanceReference": Any<String>,
  "maxContextLength": 32768,
  "modelKey": Any<String>,
  "paramsString": "3B",
  "path": "lmstudio-community/Qwen2.5-3B-Instruct-GGUF/Qwen2.5-3B-Instruct-Q4_K_M.gguf",
  "sizeBytes": 1929903008,
  "trainedForToolUse": true,
  "type": "llm",
  "vision": false,
}
`;

exports[`LLM + speculative decoding should work with .respond 1`] = `"Assistant message 2"`;

exports[`LLM + speculative decoding should work with .respond 2`] = `
{
  "acceptedDraftTokensCount": 4,
  "ignoredDraftTokensCount": 0,
  "numGpuLayers": Any<Number>,
  "predictedTokensCount": 5,
  "promptTokensCount": 41,
  "rejectedDraftTokensCount": 3,
  "stopReason": "eosFound",
  "timeToFirstTokenSec": Any<Number>,
  "tokensPerSecond": Any<Number>,
  "totalDraftTokensCount": 7,
  "totalTokensCount": 46,
  "usedDraftModelKey": Any<String>,
}
`;

exports[`LLM + speculative decoding should work with .respond 3`] = `
{
  "architecture": "qwen2",
  "contextLength": 4096,
  "displayName": "Qwen2.5 3B Instruct",
  "format": "gguf",
  "identifier": Any<String>,
  "instanceReference": Any<String>,
  "maxContextLength": 32768,
  "modelKey": Any<String>,
  "paramsString": "3B",
  "path": "lmstudio-community/Qwen2.5-3B-Instruct-GGUF/Qwen2.5-3B-Instruct-Q4_K_M.gguf",
  "sizeBytes": 1929903008,
  "trainedForToolUse": true,
  "type": "llm",
  "vision": false,
}
`;
